src/pybind/mgr/volumes/fs/operations/trash.py

```diff
@@ -53,37 +54,139 @@ def get_trash_entry(self, exclude_list):
+            # stack needed for traversing file heirarchy under trashpath in
+            # depth-first, non-recursive fashion. each stack member is an
+            # instance of class Dir (declared above, inside this method).
+            stack = deque([Dir(b"."),])
+
+            while stack and not should_cancel():
+                curr_dir = stack[-1]
+
+                # assuming True for now, if it's not empty upcoming loop will
+                # set it to False.
+                is_curr_dir_empty = True
+
+                # de = directory entry
+                de = self.fs.readdir(curr_dir.handle)
+                while de and not should_cancel():
+                    if de.d_name in (b'.', b'..'):
+                        de = self.fs.readdir(curr_dir.handle)
+                        continue
+
+                    is_curr_dir_empty = False
+                    if de.d_name in curr_dir.de_ignore_list:
+                        continue
+
+                    de_path = os.path.join(trashpath, curr_dir.path, de.d_name)
+                    if de.is_dir():
+                        try:
+                            self.fs.rmdir(de_path)
+                        except cephfs.ObjectNotEmpty:
+                            try:
+                                stack.append(Dir(os.path.join(curr_dir.path, de.d_name)))
+                                break
+                            except cephfs.Error as e:
+                                log.error(f'Exception occured: "{e}"')
+                                curr_dir.add_to_de_ignore_list(de.d_name)
+                        except cephfs.Error as e:
+                            log.error(f'Exception occured: "{e}"')
+                            curr_dir.add_to_de_ignore_list(de.d_name)
+                    else:
+                        try:
+                            self.fs.unlink(de_path)
+                        except cephfs.Error as e:
+                            log.error(f'Exception occured: "{e}"')
+                            curr_dir.add_to_de_ignore_list(de.d_name)
+
+                    try:
+                        de = self.fs.readdir(curr_dir.handle)
+                    except cephfs.Error as e:
+                        # this is the tricky one: it's an error on this
+                        # directory, not a d_name in this directory
+                        log.error(f'Exception occured: "{e}"')
+                        curr_dir.set_readdir_error()
+
+                if is_curr_dir_empty:
+                    if curr_dir.has_readdir_failed():
+                        # notify parent dir
+                        try:
+                            stack[-2].add_to_de_ignore_list(curr_dir.d_name)
+                        except IndexError:
+                            pass
+
+                    stack.pop()
+                else:
+                    curr_dir.handle.rewinddir()
+
+            self.fs.rmdir(trashpath)
+        except cephfs.ObjectNotFound:
```

Comment:
`ObjectNotEmpty`?

Reply:
> In fact, if you unlink and get that error then I would consider that success, no?

I think so too but this exception was already present in earlier version, so I wasn't sure if it should if removed. And if it should, whether it should be removed on this PR.

Reply:
> I would say you should ignore it in every place we rmdir/unlink.

Right now we are doing that since we simply return if that exception occurs -

```
        except cephfs.ObjectNotFound:
            return
```